test/e2e/deployer.go

```diff
@@ -0,0 +1,354 @@
+package e2e
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"reflect"
+	"time"
+
+	workclientset "github.com/open-cluster-management/api/client/work/clientset/versioned"
+	"github.com/open-cluster-management/work/test/e2e/bindata"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer/json"
+	"k8s.io/apimachinery/pkg/runtime/serializer/streaming"
+	"k8s.io/apimachinery/pkg/runtime/serializer/yaml"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+)
+
+type workAgentDeployer interface {
+	Deploy() error
+	Undeploy() error
+}
+
+type defaultWorkAgentDeployer struct {
+	componentNamespace string
+	clusterName        string
+	nameSuffix         string
+	image              string
+	spokeKubeClient    kubernetes.Interface
+	spokeDynamicClient dynamic.Interface
+	hubWorkClient      workclientset.Interface
+
+	clusterRoleName         string
+	clusterRoleBindingNames []string
+}
+
+func newDefaultWorkAgentDeployer(
+	componentNamespace string,
+	clusterName string,
+	nameSuffix string,
+	image string,
+	spokeKubeClient kubernetes.Interface,
+	spokeDynamicClient dynamic.Interface,
+	hubWorkClient workclientset.Interface) workAgentDeployer {
+	return &defaultWorkAgentDeployer{
+		componentNamespace: componentNamespace,
+		clusterName:        clusterName,
+		nameSuffix:         nameSuffix,
+		image:              image,
+		spokeKubeClient:    spokeKubeClient,
+		spokeDynamicClient: spokeDynamicClient,
+		hubWorkClient:      hubWorkClient,
+	}
+}
+
+func (d *defaultWorkAgentDeployer) Deploy() error {
+	// create component namespace
+	ns := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: d.componentNamespace,
+		},
+	}
+	_, err := d.spokeKubeClient.CoreV1().Namespaces().Create(context.TODO(), ns, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+
+	// create hub kubeconfig secret
+	data, err := d.getHubKubeconfigSecretData()
+	if err != nil {
+		return err
+	}
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: d.componentNamespace,
+			Name:      "hub-kubeconfig-secret",
+		},
+		Data: data,
+	}
+	_, err = d.spokeKubeClient.CoreV1().Secrets(secret.Namespace).Create(context.TODO(), secret, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+
+	// create cluster role
+	clusterRole, err := assetToUnstructured("deploy/spoke/clusterrole.yaml")
```

Comment:
I would prefer using something something like the `resourceapply.ApplyDirectly` and listing them filenames.  You'd be able to wrap them in the name setter.

Reply:
Code refactored. After discussion with qiujian16, we decided to use fixed name for resources, like component namespace cluster role and cluster rolebindings, when deploying work agent for e2e testing, because there is no conflicts between two runs of e2e test.